### PR TITLE
Ability to save shapes

### PR DESCRIPTION
**Pull request by rahul003**
### Description of changes:

Redo of PR https://github.com/awslabs/sagemaker-debugger/pull/328 rebased on current master

Updated docs as well now. 

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smdebug/tensorflow/base_hook.py [hunk 1]
```diff
@@ -306,12 +306,12 @@ def _get_writers(self, tensor_name, tensor_ref, shape_writers=False) -> List[Fil
                     else:
                         if shape_writers is False:
                             return [self.writer_map[self.device_map[self.chief_worker]]]
-                        else:
```
Comment (NihalHarish) [inline smdebug/tensorflow/base_hook.py]: Why do we need a different shape writer?
If `self._saving_shapes_in_step()` is True for every step, do we end up using only shape_writer?
Does it not make sense then to have a single writer and move the is_saving_shape logic to the Writer class?

File: smdebug/tensorflow/base_hook.py [hunk 2]
```diff
@@ -298,17 +299,25 @@ def _get_writers(self, tensor_name, tensor_ref) -> List[FileWriter]:
...
+                        if shape_writers is False:
+                            return [self.writer_map[self.device_map[self.chief_worker]]]
+                        elif self._saving_shapes_in_step():
+                            return [self.shape_writer_map[self.device_map[self.chief_worker]]]
                 elif self.save_all_workers or worker == self.chief_worker:
-                    return [self.writer_map[self.device_map[worker]]]
-            elif self.writer:
+                    if shape_writers is False:
+                        return [self.writer_map[self.device_map[worker]]]
+                    elif self._saving_shapes_in_step():
```
Comment (NihalHarish) [inline smdebug/tensorflow/base_hook.py:314]: Why do we need a different shape writer?
If self._saving_shapes_in_step() is True for every step, do we end up using only shape_writer?
Does it not make sense then to have a single writer and move the is_saving_shape logic to the Writer class?

**Review by NihalHarish — CHANGES_REQUESTED** [review]

File: smdebug/tensorflow/base_hook.py (hunk 2 above)
Comment (rahul003) [reply smdebug/tensorflow/base_hook.py]: Simplified it by merging it with writer

**Review by NihalHarish — CHANGES_REQUESTED** [review]

File: docs/analysis.md [hunk 3]
```diff
@@ -372,19 +402,19 @@ trial.tensor(name).reduction_values(step_num, mode=modes.GLOBAL, worker=None)
 ###### Returns
 `dict[(str, bool) -> numpy.ndarray]` A dictionary with keys being tuples of the form `(reduction_name, abs)` to a 1x1 numpy ndarray value. `abs` here is a boolean that denotes whether the reduction was performed on the absolute value of the tensor or not. Note that this method only returns the reductions which were saved from the training job. It does not compute all known reductions and return them if only the raw tensor was saved.
 
-#### values
-Get the values of the tensor for all steps of a given mode.
+#### shapes
```
[the remark this answers is not in the document]
Comment (rahul003) [reply docs/analysis.md:405]: This diff looks weird. I copied values contents and modified it become shapes. And moved them to group them appropriately. I didnt remove anything

File: smdebug/core/hook.py [hunk 4]
```diff
@@ -470,8 +472,12 @@ def _initialize_writers(self, only_initialize_if_missing=False) -> None:
         if self.save_all_workers is False:
             if self.worker != self.chief_worker:
                 return
+
         self.writer = FileWriter(trial_dir=self.out_dir, step=self.step, worker=self.worker)
 
+    def _get_single_process_writers(self) -> List[FileWriter]:
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/hook.py:478]: Made it _get_main_writer

File: smdebug/core/hook.py [hunk 5]
```diff
@@ -749,6 +755,28 @@ def _write_raw_tensor(self, tensor_name, tensor_value, save_collections, tensor_
                 self._write_raw_tensor_simple(tensor_name, tensor_value, tensor_ref=tensor_ref)
                 break
 
+    def _write_shape(self, tensor_name, tensor_value, save_collections, tensor_ref=None):
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/hook.py:758]: tensor value is a framework data format, it's not a tuple here. It becomes a tuple in this function.

File: smdebug/core/hook.py [hunk 6]
```diff
@@ -749,6 +755,28 @@ def _write_raw_tensor(self, tensor_name, tensor_value, save_collections, tensor_
...
+        writers = self._get_writers(tensor_name, tensor_ref=tensor_ref)
+        for s_col in save_collections:
+            reduction_config = s_col.reduction_config
+            if self.dry_run is False and reduction_config.save_shape is True:
+                numpy_tensor_value = self._make_numpy_array(tensor_value)
+                this_size, this_shape = size_and_shape(numpy_tensor_value)
+                if tensor_ref is not None and tensor_ref.tf_obj is not None:
+                    original_name = tensor_ref.tf_obj.name
+                else:
+                    original_name = None
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/hook.py:768]: Added

File: smdebug/core/index_reader.py [hunk 7]
```diff
@@ -233,28 +245,41 @@ def _update_tensors_from_json(
...
-            tensor_name = tensor["tensorname"]
-            start_idx = tensor["start_idx"]
-            length = tensor["length"]
-            tensor_location = TensorLocation(
-                tensor_name, mode, mode_step, event_file_name, start_idx, length, worker
-            )
+
+        to_update_index_dict = []
+
+        if "tensor_payload" in index_dict and len(index_dict["tensor_payload"]):
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/index_reader.py:251]: There can be empty payload

File: smdebug/core/index_reader.py [hunk 8]
```diff
@@ -233,28 +245,41 @@ def _update_tensors_from_json(
...
                 else:
-                    index_tensors_dict[tensor_name].update(
-                        {step: {worker: {"tensor_location": tensor_location}}}
-                    )
+                    index_tensors_dict[tensor_name].update({step: {worker: obj_dict}})
             else:
-                index_tensors_dict[tensor_name] = {
-                    step: {worker: {"tensor_location": tensor_location}}
-                }
+                index_tensors_dict[tensor_name] = {step: {worker: obj_dict}}
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/index_reader.py:282]: Just made the innermost dict a variable

File: smdebug/core/reduction_config.py [hunk 9]
```diff
@@ -116,7 +132,6 @@ def from_json(cls, json_str: str) -> "ReductionConfig":
         return cls.from_dict(d)
 
     def to_json_dict(self) -> Dict[str, Any]:
-        save_raw_tensor = self.save_raw_tensor
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/reduction_config.py:119]: It's just creating a copy of the variable, removed that copy. Behavior is unchanged

File: smdebug/core/tfevent/index_file_writer.py [hunk 10]
```diff
@@ -28,14 +29,21 @@ def _init_writer(self):
     def add_index(self, tensorlocation):
         if not self.writer:
             self._init_writer()
-        if not self.index_meta:
+        if not self.index_meta or not "event_file_name" in self.index_meta:
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/tfevent/index_file_writer.py:32]: Add shape can be called in some cases when add_index is never called (no tensor was saved, but only shapes). If only shapes we don't have event file. So we need this check.

File: smdebug/core/tensor.py [hunk 11]
```diff
@@ -264,6 +279,16 @@ def value(self, step_num, mode=ModeKeys.GLOBAL, worker=None):
             has_reductions = has_reduction_locations or has_reduction_values
             raise TensorUnavailableForStep(self.name, step_num, mode, has_reductions)
 
+    def shape(self, step_num, mode=ModeKeys.GLOBAL, worker=None):
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/tensor.py:282]: This is same as existing step and value methods

[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/tensor.py:282]:
```
if worker is None:
                workers = sorted(step_dict.keys())
                assert len(workers) > 0
                worker = workers[0]
```

File: smdebug/core/tensor.py [hunk 12]
```diff
@@ -126,6 +138,9 @@ class Tensor:
     def __init__(self, name, trial, cache):
         self._mode_steps = {}
         self.name = name
+        # SMdebug modifies some names of tensors to be more descriptive
```
[the remark this answers is not in the document]
Comment (rahul003) [reply smdebug/core/tensor.py:141]: Added

**Review by NihalHarish — APPROVED** [review]

**Comment by rahul003** [comment]
Merging it after Nihal's approval. If anyone has any issues, please let me know I can address them in a future PR

**Pull request merged**